libparsers/tests/csv_rw_route/preprocessor.cpp

```diff
@@ -0,0 +1,139 @@
+}
+
+TEST_CASE(
+    "libparsers - csv_rw_route - preprocessor - incorrect $Chr() expression should return "
+    "nothing") {
+	std::string test_command = "$Cr(75),$Ch(69),$Chr(75)"s;
+	parsers::errors::multi_error_t output_errors;
+	auto processed = setup(test_command, output_errors);
+	CHECK_EQ(processed.lines.size(), 1);
+	std::string value{};
+	for (auto& chr : processed.lines) {
+		value += chr.contents;
+	}
+	CHECK_EQ(value, "K");
+}
+
+TEST_CASE("libparsers - csv_rw_route - preprocessor - $Rnd") {
+	std::string test_command = "$Rnd(3;5),$Rnd(-100;100),$Rnd(0;0),$Rnd(5;3)"s;
+	parsers::errors::multi_error_t output_errors;
+	auto processed = setup(test_command, output_errors);
+	REQUIRE_EQ(processed.lines.size(), 4);
+	CHECK_LE(processed.lines[0].contents, "5"s);
+	CHECK_GE(processed.lines[0].contents, "3"s);
+
+	CHECK_LE(processed.lines[1].contents, "100"s);
+	CHECK_GE(processed.lines[1].contents, "-100"s);
+
+	CHECK_EQ(processed.lines[2].contents, "0"s);
+
+	CHECK_LE(processed.lines[3].contents, "5"s);
+	CHECK_GE(processed.lines[3].contents, "3"s);
+}
+
+TEST_CASE("libparsers - csv_rw_route - preprocessor - $Rnd bad inputs should add errors") {
+	std::string test_command = "$Rd(3;5),$nd(-100;100),$d(0;0),$Rnd(;3),$Rnd(0),$Rnd(s:a)"s;
+	parsers::errors::multi_error_t output_errors;
+	auto processed = setup(test_command, output_errors);
+	CHECK_EQ(processed.lines.size(), 0);
+	CHECK_EQ(output_errors["directive.csv"].size(), 6);
+}
+
+TEST_CASE("libparsers - csv_rw_route - preprocessor - $Sub") {
+	std::string test_command = "$Sub(0) = $Rnd(3;5)\n$Sub(1) = $Chr(75) \n,$Sub(0),$Sub(1)"s;
+	parsers::errors::multi_error_t output_errors;
+	auto processed = setup(test_command, output_errors);
+	REQUIRE_EQ(processed.lines.size(), 2);
+	CHECK_GE(processed.lines[0].contents, "3"s);
+	CHECK_LE(processed.lines[0].contents, "5"s);
+
+	CHECK_EQ(processed.lines[1].contents, "K"s);
+}
+
+TEST_CASE("libparsers - csv_rw_route - preprocessor - $If") {
+	parsers::errors::multi_error_t output_errors;
+	// Single sub expression.
+	auto processed = setup("$If(1),$Chr(75),$EndIf()"s, output_errors);
+	REQUIRE_EQ(processed.lines.size(), 1);
+	CHECK_EQ(processed.lines[0].contents, "K"s);
+
+	processed = setup("$If(0),$Chr(75),$Else(),$Chr(69),$EndIf()"s, output_errors);
+	REQUIRE_EQ(processed.lines.size(), 1);
+	CHECK_EQ(processed.lines[0].contents, "E");
+}
+TEST_CASE("libparser - csv_rw_route - preprocessor - $If multiple sub expressions") {
+	parsers::errors::multi_error_t output_errors;
+	auto processed =
+	    setup("$If(1),$Chr(75),$Chr(69),$Chr(86),$Chr(73),$Chr(78),$Else(),$Chr(69),$EndIf()"s,
+	          output_errors);
+	REQUIRE_EQ(processed.lines.size(), 5);
+	std::string value{};
+	for (auto const& line : processed.lines) {
+		value += line.contents;
+	}
+	CHECK_EQ(value, "KEVIN"s);
+
+	processed =
+	    setup("$If(0),$Chr(75),$Else(),$Chr(75),$Chr(69),$Chr(86),$EndIf()"s, output_errors);
+	REQUIRE_EQ(processed.lines.size(), 3);
+	value = ""s;
```

Comment:
Use the default constructor for a string: eg. `std::string value;`

Reply:
Not sure what you mean? I would get a redefinition error.

Reply:
Oh, derp. You should use std::string::clear() then.